File: 2021/golang/solutions/day_1.go
package solutions

import (
	utils "aoc/utils"
	"fmt"
	"log"
)

func init() {
	Solutions[1] = Day1
}

func sumWindow(values []int, windowSize int) []int {
	sums := make([]int, 0)

	for i := 0; i < len(values)-2; i++ {
		sum := 0
		for s := 0; s < windowSize; s++ {
			sum = sum + values[i+s]
		}
		sums = append(sums, sum)
	}

	return sums
}

func computeIncreases(values []int) int {
	increases := 0

	previous := values[0]
	for i := 1; i < len(values); i++ {
		current := values[i]
		if previous < current {
			increases += 1
		}
		previous = current
	}

	return increases
}

func Day1() {
	values, err := utils.ReadLinesInt("1")

	if err != nil {
		log.Fatal(err)
	}

	day1PartA(values)
	day1PartB(values)
}

func day1PartA(values []int) {
	fmt.Printf("Day 1, Part A: %v\n", computeIncreases(values))
}

func day1PartB(values []int) {
	fmt.Printf("Day 1, Part B: %v\n", computeIncreases(sumWindow(values, 3)))
}

File: 2021/elixir/lib/solutions/day1.ex
defmodule AOC.Solutions.Day1 do
  import Reader

  def partA() do
    lines_int("1")
    |> Stream.chunk_every(2, 1, :discard)
    |> Stream.filter(fn [prev, current] -> prev < current end)
    |> Enum.count()
  end

  def partB() do
    lines_int("1")
    |> Stream.chunk_every(3, 1, :discard)
    |> Stream.map(&Enum.sum/1)
    |> Stream.chunk_every(2, 1, :discard)
    |> Stream.filter(fn [prev, current] -> prev < current end)
    |> Enum.count()
  end

  def run() do
    IO.puts("Day 1, Part A: #{partA()}")
    IO.puts("Day 1, Part B: #{partB()}")
  end
end
